cluster-controller/src/main/java/io/strimzi/controller/cluster/operator/resource/StatefulSetOperator.java

```diff
@@ -251,4 +253,38 @@ public String getPodName(StatefulSet desired, int podId) {
         log.debug("Patched {} {}/{}", resourceKind, namespace, name);
         return Future.succeededFuture(ReconcileResult.patched(null));
     }
+
+    /**
+     * Reverts the changes done storage configuration of running cluster. Such changes are not allowed.
+     *
+     * @param current Current StatefulSet
+     * @param desired New StatefulSet
+     */
+    protected void revertStorageChanges(StatefulSet current, StatefulSet desired) {
+        if (current.getSpec().getVolumeClaimTemplates().isEmpty()) {
+            // We are on ephemeral storage and changing to persistent
+            desired.getSpec().setVolumeClaimTemplates(current.getSpec().getVolumeClaimTemplates());
+
+            List<Volume> volumes = current.getSpec().getTemplate().getSpec().getVolumes();
+            for (int i = 0; i < volumes.size(); i++) {
+                Volume vol = volumes.get(i);
+                if (AbstractModel.VOLUME_NAME.equals(vol.getName()) && vol.getEmptyDir() != null) {
+                    desired.getSpec().getTemplate().getSpec().getVolumes().add(0, volumes.get(i));
+                    break;
```

Comment:
Why can't we just replace the desired volume with volume from current that matches `AbstractModel.VOLUME_NAME`?

Reply:
I'm not sure I got this comment right ... but I will try ...

The `emptyDir` volumes lives in `.getSpec().getTemplate().getSpec().getVolumes()`. But the Persistent Volume doesn't. 

So basically, the `Volume` list for `ephemeral` cluster contains two volumes: `EmptyDir` volume and metrics ConfigMap volume. The `Volume` list for persistent contains only the ConfigMap with metrics only. The metrics ConfigMap might be changing, so we cannot simply overwrite it in desired. That is why we have to find the exact volume and insert it. And I have to put it in in the right order because the order matters to the diff.

Reply:
OK, got it now.